### PR TITLE
vault: 0.6.5 -> 0.7.3 with service

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

fork of PR #26130 with 
* updated to version 0.7.3
* run service under an unprivileged user instead of root
* <s>generate self-signed TLS certificate if certificate file does exist when the service starts (so ```services.vault.enable = true;``` in ```configuration.nix``` is enough to get some minimal setup up and running)</s> feature removed, see comments
* removed rarely used ```tls_*``` options (they could be set in ```extraConfig```)
* fixed typo (inMemory -> inmem)
* option names made camelCased
* build as a Makefile project, not a Go package

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/security/vault.nix [hunk 1]
```diff
@@ -1,207 +1,132 @@
...
-        { ExecStart =
-            "${pkgs.vault}/bin/vault server -config ${configFile}";
-          KillMode = "process";
-        };
-    };   
+      wantedBy = ["multi-user.target"];
+      after = [ "network.target" ];
+
+      preStart = ''
+        mkdir -m 0755 -p /var/lib/vault
```
Comment (volth) [inline nixos/modules/services/security/vault.nix]: I am not sure if this directory is used somehow, it might be a rudiment of the original PR.

Comment (copumpkin) [inline nixos/modules/services/security/vault.nix]: Does anything end up in it during normal execution?

File: nixos/modules/misc/ids.nix [hunk 2]
```diff
@@ -414,6 +415,7 @@
       btsync = 113;
       #minecraft = 114; # unused
       #monetdb = 115; # unused (not packaged), removed 2016-09-19
+      vault = 115;
```
Comment (Mic92) [inline nixos/modules/misc/ids.nix]: It looks like the service will change permission of all files related to this service on start. It might be not necessary to allocate static uids/gids.

File: nixos/modules/services/security/vault.nix (hunk 1 above)
Comment (volth) [inline nixos/modules/services/security/vault.nix]: No. But I have not tested all the features

File: nixos/modules/misc/ids.nix (hunk 2 above)
Comment (volth) [inline nixos/modules/misc/ids.nix]: It is not clear about the files, I have seen no one yet.

Comment (joachifm) [inline nixos/modules/misc/ids.nix]: If it does not even create files, a static uid is certainly unnecessary.

Comment (volth) [inline nixos/modules/misc/ids.nix]: I found out, it is for "file" storage backend, the directory name is configurable and might be different than ```/mnt/lib/vault```.
So the logic is changed: https://github.com/NixOS/nixpkgs/pull/26907/commits/09dd5ec2dc79e3183803943e4a3947bbcf9ad2c8

Anyway, we need someone experienced with Vault and security in general to review. It is my first attempt to install Vault.

Comment (joachifm) [inline nixos/modules/misc/ids.nix]: But is there a point to using a static id *and* adjusting the ownership? I can think of some, but that'd depend on the kind of data generated.

Comment (volth) [inline nixos/modules/misc/ids.nix]: adjusting of ownership is removed (it was just a migration hack as the service user was downpriveleged from root to vault)

File: nixos/modules/services/security/vault.nix [hunk 3]
```diff
@@ -92,10 +95,18 @@ in
...
       after = [ "network.target" ];
 
-      preStart = ''
-        mkdir -m 0755 -p /var/lib/vault
-        chown -R vault:vault /var/lib/vault
-
+      preStart =
+        optionalString (cfg.storageBackend == "file" || cfg.storageBackend == "file_transactional")
+          (let
+            matched = builtins.match ''.*path[ ]*=[ ]*"([^"]+)".*'' (toString cfg.storageConfig);
```
Comment (volth) [inline nixos/modules/services/security/vault.nix]: in nix 1.12 there is no ```\s*``` nor ``` *``` in regex :(

File: nixos/modules/misc/ids.nix (hunk 2 above)
Comment (volth) [inline nixos/modules/misc/ids.nix]: I am not sure if ```vault``` _group_ is needed

File: nixos/modules/services/security/vault.nix [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    };
+  };
+
+  config = let
+    localDir = if (cfg.storageBackend == "file" || cfg.storageBackend == "file_transactional") then
+                 let
+                   matched = builtins.match ''.*path[ ]*=[ ]*"([^"]+)".*'' (toString cfg.storageConfig);
+                 in
+                   if matched == null then
+                     throw ''`storageBackend` "${cfg.storageBackend}" requires path in `storageConfig`''
```
Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: This error message is going to bring tears to someone someday.

File: nixos/modules/services/security/vault.nix [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+           ++ optional (config.services.consul.enable && cfg.storageBackend == "consul") "consul.service";
+
+      preStart = optionalString (localDir != null) ''
+        install -d -m0700 -o vault -g vault "${localDir}"
+      '' + ''
+        # generate a self-signed certificate, you will have to set environment variable "VAULT_SKIP_VERIFY=1" in the client
+        if [ ! -s ${cfg.tlsCertFile} -o ! -s ${cfg.tlsKeyFile} ]; then
+          mkdir -p $(dirname ${cfg.tlsCertFile}) || true
+          mkdir -p $(dirname ${cfg.tlsKeyFile }) || true
+          ${pkgs.openssl.bin}/bin/openssl req -x509 -newkey rsa:2048 -sha256 -nodes -days 99999 \
```
Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: These constants (including the algorithm) should be configurable.

Comment (volth) [inline nixos/modules/services/security/vault.nix]: The "normal" use case is provide own ```tlsCertFile``` and ```tlsKeyFile``` files, signed with trusted CA.
Self-signed certificate is just a bit more secure replacement for turning TLS off when there is no production-ready certificate.

Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: Unconvincing and incidentally not true.

Comment (volth) [inline nixos/modules/services/security/vault.nix]: Maybe better to remove this cert generation from nixpkgs to avoid confusions.

Comment (volth) [inline nixos/modules/services/security/vault.nix]: There is only problem: vault's console client works only via HTTPS.
So I would prefer to keep self-signing as default, for easy deployment.

**Comment by volth** [comment]
Vault actually a Makefile project, not a Go package.
It works being compiled as a Go package, but some functions fail
```
$ vault version   
Vault (version unknown) (cgo)
```

File: nixos/modules/services/security/vault.nix [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        default = "";
+        description = "Telemetry configuration";
+      };
+    };
+  };
+
+  config = let
+    localDir = if (cfg.storageBackend == "file" || cfg.storageBackend == "file_transactional") then
+                 let
+                   matched = builtins.match ''.*path[ ]*=[ ]*"([^"]+)".*'' (toString cfg.storageConfig);
```
Comment (cstrahan) [inline nixos/modules/services/security/vault.nix]: Instead of trying to parse the option, how about this: rename `storageConfig` to `storagePath`, and then in the config file, prefix as necessary (i.e. `path = ${cfg.storagePath}`). That will also allow you to remove the `throw` below, and instead use something like:

```
    assertions = [
      { assertion = lib.elem cfg.storageBackend [ "file" "file_transactional" ] && cfg.storagePath == null;
        message = ''You must set services.vault.storagePath when using either the "file" or "file_transactional" backend'';
      }
    ]
```

My understanding is that `assertions` are aggregated, so rather than causing the interpreter to bail out on the first `throw`, the user should get a much more pleasant experience with a list of the failed assertions.

Comment (volth) [inline nixos/modules/services/security/vault.nix]: good point, thank you!

Comment (volth) [inline nixos/modules/services/security/vault.nix]: ```storageConfig``` is needed too, for the "consul" and database backends (then it contains address and credentials)

Comment (cstrahan) [inline nixos/modules/services/security/vault.nix]: Good point regarding `storageConfig` - this is looking really nice!

**Comment by volth** [comment]
@cstrahan, ok

File: pkgs/tools/security/vault/default.nix [hunk 7]
```diff
@@ -7,27 +7,32 @@ let
     rev = "e2f59b64be1fa5430fa05c91b6274284de4ea77c";
     sha256 = "10m75rp3hy71wlmnd88grmpjhqy0pwb9m8wm19l0f463xla54frd";
   };
-in buildGoPackage rec {
+in stdenv.mkDerivation rec {
```
Comment (copumpkin) [inline pkgs/tools/security/vault/default.nix]: Not sure if I missed the comment, but why the switch away from `buildGoPackage`? I'm not saying you shouldn't, but I'd want to make sure any annoyances you run into with `buildGoPackage` get back to @kamilchm and get improved over time.

Comment (volth) [inline pkgs/tools/security/vault/default.nix]: Unlike a typical Go project, Vault sources have a Makefile.

Vault works being compiled with ```buildGoPackage```, but some functions fail, for example:
```
$ vault version   
Vault (version unknown) (cgo)
```

Comment (Mic92) [inline pkgs/tools/security/vault/default.nix]: Switching from buildGoPackage to mkDerivation has the disadvantage, that will contain a runtime reference on the go compiler itself. This is usually cleared by buildGoPackage.

Comment (0xABAB) [inline pkgs/tools/security/vault/default.nix]: I'd like to see `vault version` report a useful version.

Comment (volth) [inline pkgs/tools/security/vault/default.nix]: @0xABAB it reports if compiled with Makefile

**Comment by volth** [comment]
one more problem: ```nixos-rebuild switch``` should not restart Vault, restart turns Vault into the "sealed" state. I will fix it tomorrow.

File: pkgs/tools/security/vault/default.nix (hunk 7 above)
Comment (kamilchm) [inline pkgs/tools/security/vault/default.nix]: It's a common pattern to set version from git with `go build -ldflags` into Go variable used to show it later https://github.com/hashicorp/vault/blob/5d2d750d9a5771d321a918d80dd0f160941fde68/scripts/build.sh#L50

We could try to extend `buildGoPackage` for configurable `ldfags` if this is the only problem.

Comment (volth) [inline pkgs/tools/security/vault/default.nix]: Another difference (not sure if it is a problem) - ```buildGoPackage``` produces 5 files in bin/, ```Makefile``` - only one.

Comment (volth) [inline pkgs/tools/security/vault/default.nix]: The git commit is lost anyway (```fetchFromGitHub``` loses it), the problem is this [version_base.go](https://github.com/hashicorp/vault/blob/5d2d750d9a5771d321a918d80dd0f160941fde68/version/version_base.go) is not compiled by ```buildGoPackage``` and ```"0.7.3"``` is lost

Comment (kamilchm) [inline pkgs/tools/security/vault/default.nix]: This can be limited by `subPackages`. Looking into `build.sh` I suspect that `subPackages = [ "." ];` should produce the same bin file as `Makefile`.

I don't try to convince you to use `buildGoPackage`, just saying there are options to make it easier.

Comment (volth) [inline pkgs/tools/security/vault/default.nix]: How to tell ```buildGoPackage``` not to exclude "version_base.go" ?

Comment (kamilchm) [inline pkgs/tools/security/vault/default.nix]: according to https://github.com/hashicorp/vault/blob/5d2d750d9a5771d321a918d80dd0f160941fde68/version/version_base.go#L1 you need to set `vault` into `-tags` using `buildFlags` like https://github.com/NixOS/nixpkgs/blob/bf5dc90f304b6dd047a031cd097ada3bc242b7e0/pkgs/tools/security/keybase/default.nix#L19

**Comment by 0xABAB** [comment]
@volth That does imply that the `autoUpgrade` feature of NixOS becomes useless.

**Comment by volth** [comment]
@0xABAB Sure. It is one of those services which cannot be upgraded automatically (```grep -R 'restartIfChanged = false' *``` for the others) because it requires the operator to enter the unseal passwords just after upgrade.

**Comment by 0xABAB** [comment]
@volth It would be useful to have a warning printed when an operator makes use of such features that are not autoupgradable. That removes some surprises and also encourages different deployment strategies. (I am not saying that you should solve this problem here, but it's more such that if someone else feels the same that they can point to this comment as evidence.)

I suppose the autoUpgrade feature is just fundamentally flawed for any non-trivial setup.

**Comment by 0xABAB** [comment]
I don't think it's a good idea to spend more time on various details of this PR. If people care about some of those details, then file issues or PRs for those.

**Comment by volth** [comment]
@0xABAB ```nixos-rebuild``` prints in CAPS about some modules were not restarted (usually there is only ```display-manager```)

**Comment by volth** [comment]
0xABAB. Those details are ones I care about, so it would be me who open the next PR if this PR accepted too quickly (anyway I have very limited influence on when it happens)

**Comment by 0xABAB** [comment]
@volth I am not going to argue against the person doing the PR. So, that's easy. Thanks for your other reply.

File: pkgs/tools/security/vault/default.nix [hunk 8]
```diff
@@ -7,27 +7,35 @@ let
...
   '';
 
-  postInstall = ''
-    mkdir -p $bin/share/bash-completion/completions/
-    cp ${vaultBashCompletions}/vault-bash-completion.sh $bin/share/bash-completion/completions/vault
+  installPhase = ''
+    mkdir -p $out/bin $out/share/bash-completion/completions
+
+    cp pkg/*/* $out/bin/
+    find $out/bin -type f -exec remove-references-to -t ${go} '{}' +
```
Comment (0xABAB) [inline pkgs/tools/security/vault/default.nix]: I would like to see double quotes used everywhere a path is referenced. (Various times in this file.)

File: nixos/modules/services/security/vault.nix [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+      enable = mkEnableOption "Vault daemon";
+
+      address = mkOption {
+        type = types.str;
+        default = "127.0.0.1:8200";
+        description = "The name of the ip interface to listen to";
+      };
+
+      tlsCertFile = mkOption {
```
Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: For `tls{Key,Cert}File` I'd like to see sane defaults defined, but they just wouldn't be active (so, nothing changes compared to the  current PR), because a new boolean option "tls.enabled" (or something more in line with the naming conventions there are now) would be introduced that by default is false. This way there is more potential standardization.

File: nixos/modules/services/security/vault.nix [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        example = "/path/to/your/key.pem";
+        description = "TLS private key file. TLS will be disabled unless this option is set";
+      };
+
+      listenerExtraConfig = mkOption {
+        type = types.lines;
+        default = ''
+          tls_min_version = "tls12"
+        '';
+        description = "extra configuration";
```
Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: Documenting what all the valid values for this option are, would be helpful.

File: pkgs/tools/security/vault/default.nix [hunk 11]
```diff
@@ -7,27 +7,35 @@ let
...
-  buildFlagsArray = ''
-    -ldflags=
-      -X github.com/hashicorp/vault/version.GitCommit=${version}
+  nativeBuildInputs = [ go gox removeReferencesTo ];
+
+  buildPhase = ''
+    substituteInPlace scripts/build.sh --replace 'git rev-parse HEAD' 'echo ${src.rev}'
+
+    mkdir -p src/github.com/hashicorp
+    ln -s $(pwd) src/github.com/hashicorp/vault
```
Comment (0xABAB) [inline pkgs/tools/security/vault/default.nix]: Quotes

File: nixos/modules/services/security/vault.nix (hunk 9 above)
Comment (volth) [inline nixos/modules/services/security/vault.nix]: Both default (```tls.enabled``` with ```false``` and self-signed certificates) are bad, because Vault client does not work with them well.
The former requires to run it as ```VAULT_ADDR=http://127.0.0.1:8200 vault status```, the latter ```VAULT_SKIP_VERIFY=1 vault status```.
I expect you'd offer to wrap the client executable according to the server setting.

File: pkgs/tools/security/vault/default.nix (hunk 8 above)
Comment (volth) [inline pkgs/tools/security/vault/default.nix]: exactly this line is copy-pasted so could be found various times in nixpkgs. Wouldn't it be better to open a separate PR to fix them all at once?

File: nixos/modules/services/security/vault.nix (hunk 9 above)
Comment (0xABAB) [inline nixos/modules/services/security/vault.nix]: I meant default *file names*.

File: pkgs/tools/security/vault/default.nix (hunk 8 above)
Comment (0xABAB) [inline pkgs/tools/security/vault/default.nix]: First people said they didn't want to change code, because others wrote it while they were working on it in and now you are saying that you don't want to change new code that you copy pasted. 

I am completely against this, but in a world where everyone is crazy this could be a good plan. 

A separate PR to fix all similar issues that came before you would also be welcome, but obviously that doesn't need to come from you.

**Comment by cstrahan** [comment]
Alright, I think this has simmered long enough. Thanks at ton, @volth !
